### PR TITLE
LLD TypeScript fix typing of families and modals

### PR DESCRIPTION
**Pull request by gre**
### 📝 Description

LLD TypeScript improvements notably including

- [x] eslint support of `{ foo: _foo, ...rest }` pattern
- [x] modals typing (MVP was https://github.com/LedgerHQ/ledger-live/pull/3429 )
- [x] coin family type establishment (MVP was https://github.com/LedgerHQ/ledger-live/pull/3371 )
- [x] families folder fixes

# Fixing all the type errors of families/

the main work that this PR delivers was to fix all errors from LLD families folder. but we did it by also reinforcing the type of what is a coin family, as well as the modals data input types.

# LLDCoinFamily contract type

- **`generated/` folder is gone:** in the process of fixing typescript of coin families folder, we remove completely this complex folder. We replace it with just one `generated.ts` which is the result of all the `families/*/index.ts` and the `families/*/modals.ts`
- each family export one **`index.ts`** and only exports a module object of type `LLDCoinFamily`.
- each family also exports their modals. that we properly typecheck locally in the family and globally among the project.
- this unified type will allow us to more easily maintain the "Coin Family interface" to implement in LLD, easily deprecate things if we have to, as well as detect typescript issues.
- this unified type is self documenting the API you have to implement with comments on each component that you can define (instead of the current status quo of copying things from another family!)

#### Better typing with parametric binding specifics' Account, Transaction and TransactionStatus

The type is reinforced in that you can define your own specific account/transaction/status and it's about the only place where you are allowed to do these (inside your family). This does not involve casting at all!

```typescript
import { CeloAccount, Transaction, TransactionStatus } from "@ledgerhq/live-common/families/celo/types";
const family: LLDCoinFamily<CeloAccount, Transaction, TransactionStatus> = {
```

this will allow you to drop any code like `invariant(account.family === "celo")` and `account as CeloAccount`, you have typesafety on this and all components will directly be able to define `({ account: CeloAccount })` as props.

let's hope this gives idea to eventually generalize the same on the bridge interface!

this essentially works by the fact the type is defined as:

```
type LLDCoinFamily<
  A extends Account,
  T extends TransactionCommon,
  TS extends TransactionStatus,
  CoinModalsData = {}
>
```

#### NO BREAKING CHANGES from a family perspective...

an important choice of this rework was to NOT break any signature and existing design choices.

#### ...BUT just one file to add, and types to fix to it:

The impact is very limited. You just need to add a index.ts file. you can start with this template and remove the files you don't have:

```typescript
import "./live-common-setup";
import {
  _FAMILY_Account,
  Transaction,
  TransactionStatus,
} from "@ledgerhq/live-common/families/_FAMILY_/types";
import { LLDCoinFamily } from "../types";
import operationDetails from "./operationDetails";
import accountActions from "./accountActions";
import accountHeaderManageActions from "./AccountHeaderManageActions";
import transactionConfirmFields from "./TransactionConfirmFields";
import AccountBodyHeader from "./AccountBodyHeader";
import AccountSubHeader from "./AccountSubHeader";
import sendAmountFields from "./SendAmountFields";
import sendRecipientFields from "./SendRecipientFields";
import sendWarning from "./SendWarning";
import receiveWarning from "./ReceiveWarning";
import AccountBalanceSummaryFooter from "./AccountBalanceSummaryFooter";
import tokenList from "./TokenList";
import StepReceiveFunds from "./StepReceiveFunds";
import StepReceiveFundsPostAlert from "./StepReceiveFundsPostAlert";
import NoAssociatedAccounts from "./NoAssociatedAccounts";
import StakeBanner from "./StakeBanner";

const family: LLDCoinFamily<_FAMILY_Account, Transaction, TransactionStatus> = {
  operationDetails,
  accountActions,
  accountHeaderManageActions,
  transactionConfirmFields,
  AccountBodyHeader,
  AccountSubHeader,
  sendAmountFields,
  sendRecipientFields,
  sendWarning,
  receiveWarning,
  AccountBalanceSummaryFooter,
  tokenList,
  StepReceiveFunds,
  StepReceiveFundsPostAlert,
  NoAssociatedAccounts,
  StakeBanner,
};

export default family;
```

you will then need fix all the TypeScript errors that become obvious when you don't implement the contract:

Screenshot 2023-05-26 at 10 31 41

**please also note there is a special convention on the file `modals.ts` that you must define if you want to export modals too.** Check other coins for reference.

# Modals rework

The modals have been properly typed.

This is done in a very "KISS" (keep it simple stupid) way:

- no enum, we still just use string, but they are keys in a global type object that define all possible modals. Each key is used to bond everything together (the modal id and the related data types).
- no breaking change
  - `openModal` have the same signature, but have now full typechecking
  - Modal still works like before but we have typechecking on the name prop that is set against the data that is received.

demo of type check & inference:

238030858-d607f181-837b-402f-a854-6f121b65462e

### ❓ Context

- **Impacted projects**: `LLD` 
- **Linked resource(s)**: https://ledgerhq.atlassian.net/browse/LIVE-4264 

### ✅ Checklist

- [ ] **Test coverage** 
- [x] **Atomic delivery** 
- [x] **No breaking changes** 

### 🚀 Expectations to reach

_Please make sure you follow these [**Important Steps**](https://github.com/LedgerHQ/ledger-live/blob/develop/CONTRIBUTING.md#important-steps)._

_Pull Requests must pass the CI and be internally validated in order to be merged._

### CODE REVIEW

File: apps/ledger-live-desktop/src/renderer/families/celo/LockFlowModal/steps/StepConfirmation.tsx [hunk 1]
```diff
@@ -1,35 +1,32 @@
...
-  const account = useSelector(s =>
-    accountSelector(s, {
-      accountId: initialAccount?.id,
-    }),
-  );
+  // const account = useSelector(s =>
+  //   accountSelector(s, {
+  //     accountId: initialAccount?.id,
+  //   }),
+  // );
```
Comment (chabroA) [inline apps/ledger-live-desktop/src/renderer/families/celo/LockFlowModal/steps/StepConfirmation.tsx:29]: This looks like dead code

```suggestion

```

**Review by chabroA — APPROVED** [review]
LGTM. Left a comment regarding dead code and it looks like the CI Mobile Build App job is failling

**Review by haammar-ledger — APPROVED** [review]
LGTM for `families/*`

File: apps/ledger-live-desktop/src/renderer/families/types.ts [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+  eventProperties?: Record<string, unknown>;
+  icon: React.ComponentType<IconType> | ((a: { size: number }) => React.ReactElement);
+  disabled?: boolean;
+  tooltip?: string;
+  accountActionsTestId?: string;
+};
+
+// the AllCoinFamilies type is the only time we accept to go "any" because it's the generated entry point where the bridge is rooted.
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export type AllCoinFamilies = Record<string, LLDCoinFamily<any, any, any>>;
```
[the remark this answers is not in the document]
Comment (gre) [reply apps/ledger-live-desktop/src/renderer/families/types.ts:311]: not even in the (not committed) generated.ts? you may be right.
@henrily-ledger  / @valpinkman can you check?

**Comment by gre** [comment]
@valpinkman could you check the CI to make sure it pass? seems we fail on desktop test, maybe tmp flawky case happened as it was earlier this week

File: apps/ledger-live-desktop/src/renderer/families/types.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (valpinkman) [reply apps/ledger-live-desktop/src/renderer/families/types.ts:311]: Yes it's used on many families
eg `families/polkadot/operationDetail.tsx`

```
import { AmountCellExtraProps, OperationDetailsExtraProps } from "../types";
```

[the remark this answers is not in the document]
Comment (gre) [reply apps/ledger-live-desktop/src/renderer/families/types.ts:311]: mmh but `AllCoinFamilies`

[the remark this answers is not in the document]
Comment (gre) [reply apps/ledger-live-desktop/src/renderer/families/types.ts:311]: no used, i will remove

**Pull request merged**